First-class function type

pearu opened the pull request:
Closes #3405 

This PR (as a continuation of PR #4967) implements:
- [x] Major revision on how Dispatcher and FunctionType instances interact.
- [x] Full support for jit-decorated functions as first-class type functions
    - The jit-decorated functions are considered first-class type functions when using `firstclass=True` option. For instance:
        ```
        @numba.njit
        def a(x):  # `a` is first-class type function, that is,
                   # it can be used as an object in other jit-decorated functions
            return x + 1

        @numba.jit()
        def foo(f, x):
            return f(x)

        assert foo(a, 3) == 4
        ```

pearu commented:
@stuartarchibald to reproduce the issue of constant functions:
1. uncomment the line in https://github.com/Quansight/numba/blob/6e190ed13987c2c33d065eebc807879e3307a9a0/numba/core/types/function.py#L116
2. run the test_constant_functions test, for instance:
    ```
    pytest -sv numba/tests/test_function_type.py -k test_constant_functions
    ```
    that should print out the LLVM IR and fail with `AssertionError: 246 != 579`

sklam commented:
It needs to handle njit functions that are generators, or reject gracefully:

```python
from numba import njit

@njit(firstclass=True)
def gen(xs):
    for x in xs:
        x += 1
        yield x

@njit
def con(gen_fn, xs):
    return  [it for it in gen_fn(xs)]


con(gen, (1, 2, 3))

```
currently, it fails with:

```pytb
numba/core/pythonapi.py in serialize_object(self, obj)
   1327         try:
-> 1328             gv = self.module.__serialized[obj]
   1329         except KeyError:

KeyError: (UniTuple(int64 x 3),) -> int64 generator(func=<function gen at 0x116defb90>, args=(UniTuple(int64 x 3),), has_finalizer=True)
```

pearu commented:
In nopython mode, generators cannot be used as first-class function types because of the corresponding cfunc wrapper functions cannot hold the state of generators. So, raising `UnsupportedError` when `njit(firstclass=True)`-decorating generators.

In python mode, however, generators can be used as first-class functions.

pearu commented:
To eliminate the need for `firstclass=True`, the notes in https://github.com/xnd-project/rbc/wiki/How-numba-resolves-function-calls%3F#numba-jit-decorated-functions-as-first-class-functions must be addressed.

Also, the implementation of `FunctionType` for jit-decorated functions needs to be made immutable: currently, the hash value depends on the last specialization of jit-decorated function.

pearu commented:
This PR is ready for review. Here are notable changes:
- Revised the interaction between FunctionType and Dispatcher instances. Introduced UndefinedFunctionType that is an imprecise Numba type during type-inference and that is transformed to be precise FunctionType during lowering.
- Enabled all tests for Numba jit-decorated functions, cfunc-decorated functions, and WAP instances.
- Removed the need for `firstclass=` jit decorator option.
- Generating cfunc wrapper functions is enabled by default for all Numba jit-decorated functions except for generator functions, and functions that take `Omitted` arguments or return `Optional` values.
- Simplified and cleaned up the code. Improved error handling from using first-class functions.

@sklam @stuartarchibald, please review.

sklam commented:
1 test is failing: numba.tests.test_cfunc.TestCFunc.test_errors [azure log](https://dev.azure.com/numba/numba/_build/results?buildId=4699&view=logs&j=a506c6d0-8428-5bb5-7d36-c242a5467a36&t=fbf6f884-d589-58cc-173a-ae041ad78914&l=415).
The test might have aborted the process.

stuartarchibald commented:
Fail details:
<details>

```
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/vsts/work/1/s/numba/runtests.py", line 9, in <module>
    sys.exit(0 if _main(sys.argv) else 1)
  File "/home/vsts/work/1/s/numba/testing/_runtests.py", line 26, in _main
    **kwds).wasSuccessful()
  File "/home/vsts/work/1/s/numba/testing/__init__.py", line 73, in run_tests
    nomultiproc=nomultiproc)
  File "/home/vsts/work/1/s/numba/testing/main.py", line 160, in __init__
    super(NumbaTestProgram, self).__init__(*args, **kwargs)
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.7/unittest/main.py", line 101, in __init__
    self.runTests()
  File "/home/vsts/work/1/s/numba/testing/main.py", line 321, in runTests
    run_tests_real()
  File "/home/vsts/work/1/s/numba/testing/main.py", line 306, in run_tests_real
    super(NumbaTestProgram, self).runTests()
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.7/unittest/main.py", line 271, in runTests
    self.result = testRunner.run(self.test)
  File "/home/vsts/work/1/s/numba/testing/main.py", line 734, in run
    return super(ParallelTestRunner, self).run(self._run_inner)
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.7/unittest/runner.py", line 176, in run
    test(result)
  File "/home/vsts/work/1/s/numba/testing/main.py", line 682, in _run_inner
    self._run_parallel_tests(result, pool, child_runner, tests)
  File "/home/vsts/work/1/s/numba/testing/main.py", line 718, in _run_parallel_tests
    raise e
  File "/home/vsts/work/1/s/numba/testing/main.py", line 709, in _run_parallel_tests
    child_result = it.__next__(self.timeout)
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.7/multiprocessing/pool.py", line 743, in next
    raise TimeoutError from None
multiprocessing.context.TimeoutError: Tests didn't finish before timeout (or crashed):
- 'numba.tests.test_cfunc.TestCFunc.test_errors'
```

</details>

stuartarchibald commented:
/AzurePipelines run

stuartarchibald commented:
Rerunning just in case it's a Frosty Thompson or similar.

stuartarchibald commented:
Failure is consistent.

sklam commented on numba/core/lowering.py:
```diff
@@ -926,6 +938,58 @@ def _lower_call_RecursiveCall(self, fnty, expr, signature):
+                cgutils.is_null(self.builder, addr),
+                likely=False) as (then, orelse):
+            with then:
+                # TODO: acquire gil??
```
Yes, it will need the GIL.

pearu replied:
Fixed.

sklam commented on numba/core/lowering.py:
```diff
@@ -926,6 +938,58 @@ def _lower_call_RecursiveCall(self, fnty, expr, signature):
+                # cgutils.printf(self.builder, "pyaddr=%p\n", pyaddr)
+                # todo use self.api?
+                pyapi = self.context.get_python_api(self.builder)
+                addr1 = lower_get_wrapper_address(
```
This is interesting. So it will callback into the interpreter and may trigger more compilation.  I wonder is it surviving caching or serialization (of the function). Also, the re-entrant into interpreter seems expensive and it's violating the `nopython=True`. (`print()` violates it but it's usually for debug only, while this feature is not debug only).

pearu replied:
Re-entering to the interpreter is triggered only when `addr==0` for some reason, most likely for some corner-case that I cannot specify atm. In fact, the condition would most likely indicate a bug.

In all unittest cases, the condition is not satisfied and the interpreter is not used.

I would be fine to remove the `lower_get_wrapper_address` call and just trigger an exception when the `addr == 0` turns to be true. What do you think?

pearu replied:
The `test_zero_address` test constructs a case (see `Bad2Good`) where the above code is fully used. Marking as resolved.

sklam commented on numba/core/types/function_type.py:
```diff
@@ -0,0 +1,200 @@
+        if sig is None:
+            return self
+
+        self._reset_signature(sig)
```
A type object should not mutate itself such that it's `.key` changes.  `.key` is used to compute the hash of the type.  A non-precise type should never become a precise type.  They should define a `.refine()` method that creates a new type that is the precise version.

pearu replied:
Fixed.
Btw, when reading the `refine` usage, I noticed:
https://github.com/numba/numba/blob/5364a294ce47dfc97628d3fe82b77441d0dab8a5/numba/core/types/containers.py#L524
that does nothing. Is this a bug and be `assert res.is_precise()` or similar?
Another case is
https://github.com/numba/numba/blob/5364a294ce47dfc97628d3fe82b77441d0dab8a5/numba/core/types/containers.py#L606

sklam replied:
right, there should be an `assert`

pearu replied:
Fixed by turning the is_precise calls into assert statements.

sklam commented:
The following is failing at lowering. It needs a more graceful error:

```python
from numba import njit

@njit
def f1(x):
    return x


@njit
def f2(x):
    return x


@njit
def foo(disp1, disp2, sel):
    if sel == 1:
        fn = disp1
    else:
        fn = disp2


    return fn([1]), fn(1)


print(foo(f1, f2, sel=1))
```

```pytb
Traceback (most recent call last):
  File "/Users/siu/dev/numba-git/numba/core/errors.py", line 720, in new_error_context
    yield
  File "/Users/siu/dev/numba-git/numba/core/lowering.py", line 266, in lower_block
    self.lower_inst(inst)
  File "/Users/siu/dev/numba-git/numba/core/lowering.py", line 347, in lower_inst
    val = self.lower_assign(ty, inst)
  File "/Users/siu/dev/numba-git/numba/core/lowering.py", line 521, in lower_assign
    return self.lower_expr(ty, value)
  File "/Users/siu/dev/numba-git/numba/core/lowering.py", line 1041, in lower_expr
    res = self.lower_call(resty, expr)
  File "/Users/siu/dev/numba-git/numba/core/lowering.py", line 778, in lower_call
    res = self._lower_call_FunctionType(fnty, expr, signature)
  File "/Users/siu/dev/numba-git/numba/core/lowering.py", line 944, in _lower_call_FunctionType
    assert fnty.check_signature(signature)
AssertionError
```

pearu commented:
The CI failure should be now fixed (it was due to uncomplete error handling code that missed gil release).

The error message from @sklam example of value-dependent polymorphism is now more informative and the corresponding test case has been added to unittests.

All review comments have been addressed. Please review.

sklam commented on numba/core/function_type.py:
```diff
@@ -0,0 +1,261 @@
+                                                info=type(pyval).__name__)
+        return sfunc._getvalue()
+
+    if isinstance(pyval, Dispatcher):
```
This branch seems unreachable.

pearu replied:
yes, removed the `if` block.

sklam commented on numba/core/function_type.py:
```diff
@@ -0,0 +1,261 @@
+    if isinstance(pyval, CFunc):
+        addr = pyval._wrapper_address
+        sfunc = cgutils.create_struct_proxy(typ)(context, builder)
+        sfunc.addr = context.add_dynamic_addr(builder, addr,
```
The use of `add_dynamic_addr` will disable caching.

pearu replied:
So it means that if a Numba jit-decorated function uses a first-class function that is passed in via namespace lookup (rather than via argument passing) then such a function cannot be cached.
Is there a way to detect if the user tries to cache such functions so that the exception could be raised at the lower_constant call?

sklam replied:
> Is there a way to detect if the user tries to cache such functions so that the exception could be raised at the lower_constant call?

Cannot do that consistently. What will happen is that the `add_dynaic_addr` would place something into the LLVM IR such that later the caching layer will raise a warning.

pearu replied:
OK. I suggest no changes since eventually there will be a warning.

sklam commented on numba/core/typed_passes.py:
```diff
@@ -361,6 +360,20 @@ def run_pass(self, state):
+                                      (types.Optional, types.Generator)):
+                            pass
+                        else:
+                            lower.create_cfunc_wrapper()
```
Need to modify the flags in the other GPU compiler pipelines to turn off `no_cfunc_wrapper`. The CUDA backend is currently failing because `CUDATargetContext` doesn't have `.create_cfunc_wrapper`.

pearu replied:
Fixed by setting `no_cfunc_wrapper=True` when not using the cpu compiler.

sklam commented on numba/core/types/function_type.py:
```diff
@@ -0,0 +1,189 @@
+        """
+
+
+class CompileResultWAP(WrapperAddressProtocol):
```
This class doesn't seem to be used at all.

pearu replied:
This is true. Although it can be useful for turning dispatcher instance compilation results into first-class functions with precise types.

sklam replied:
To keep this, we need to have a test that exercises it and have docstrings.

pearu replied:
Added a new test `test_compilation_results` and docstrings to `CompileResultWAP`.

sklam commented on numba/core/utils.py:
```diff
@@ -449,3 +450,134 @@ def chain_exception(new_exc, old_exc):
+            yield (t, ) + rest
+
+
+def iter_multivalued_args(args):
```
Unused?

pearu replied:
yes, it can be removed.

pearu replied:
Removed.

sklam commented on numba/core/utils.py:
```diff
@@ -449,3 +450,134 @@ def chain_exception(new_exc, old_exc):
+    return min_nargs, max_nargs
+
+
+def resolve_dispatcher_types(numba_types):
```
unused?

pearu replied:
yes, it can be removed.

pearu replied:
Removed.

sklam commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,118 @@ the beginning or the end of the index specification::
+as items in sequences, in addition of being callable.
+
+To enable first-class function support, Numba :term:`JIT` compiled
+functions must specify ``firstclass=True`` ``jit`` option. Numba
```
The docs need to be updated. `firstclass` option is removed in this PR.

pearu replied:
Fixed.

sklam commented on numba/core/function_type.py:
```diff
@@ -0,0 +1,261 @@
+
+def lower_get_wrapper_address(context, builder, func, sig,
+                              failure_mode='return_exc'):
+    """Low-level call to _get_wrapper_address(func, sig)
```
Can you add a note here that this function needs the GIL acquired please?

pearu replied:
Yes, added: "When calling this function, GIL must be acquired."

sklam commented on numba/core/function_type.py:
```diff
@@ -0,0 +1,261 @@
+    pyapi = context.get_python_api(builder)
+    sfunc = cgutils.create_struct_proxy(toty)(context, builder)
+
+    addr = lower_get_wrapper_address(
```
This function needs the GIL.

pearu replied:
Fixed.

sklam commented on numba/core/function_type.py:
```diff
@@ -0,0 +1,261 @@
+        context, builder, val, toty.signature,
+        failure_mode='return_exc')
+    sfunc.addr = pyapi.long_as_voidptr(addr)
+    pyapi.decref(addr)
```
decref requires the GIL

pearu replied:
Fixed.

@sklam when calling `lower_get_wrapper_address` from `unbox`, does it need to acquire GIL as well?

sklam replied:
`unbox` is executed with the GIL still acquired. So no.

sklam commented on numba/core/lowering.py:
```diff
@@ -926,6 +938,65 @@ def _lower_call_RecursiveCall(self, fnty, expr, signature):
+                cgutils.is_null(self.builder, addr),
+                likely=False) as (then, orelse):
+            with then:
+                self.init_pyapi()
```
Is this branch being exercised by any test? I tried to cause this branch to error out (i.e. storing NULL as the function pointer) but none of the test would fail.

pearu replied:
No, addr should be never be zero here.
This branch is exercised in the case of some bug and it is implemented in order to fail gracefully with a RuntimeError exception rather than a segfault.

sklam replied:
I meant this branch in the `with then`. Precisely the segment https://github.com/Quansight/numba/blob/6d361f7a78c86725e7307ba83269de26a01fae58/numba/core/lowering.py#L971-L995. I cannot find a test that is executing the compiled code. So I tried to set the address `addr2` (at https://github.com/Quansight/numba/blob/6d361f7a78c86725e7307ba83269de26a01fae58/numba/core/lowering.py#L993) to zero to see if a test would fail, but no test failed. I think a new test is needed to exercise this segment.

pearu replied:
Added a new test that exercises the above-mentioned branch.
See test_zero_address in test_function_type, in particular, the Bad2Good example.

pearu commented:
@sklam I have addressed all your comments and feedback. Please review.

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+- the decorated function has Omitted arguments,
+- or the decorated function returns Optional value.
+
+To disable first class function support, use `no_cfunc_wrapper=True`
```
```suggestion
To disable first class function support, use ``no_cfunc_wrapper=True``
```

pearu replied:
Fixed.

stuartarchibald commented:
@sklam some notes from our discussion.

* Need to move import from `numba/__init__.py` to somewhere more appropriate.
* Need to move any reasonable candidates to `numba/experimental/`.
* Issue warning of experimental feature if used (i.e. unification of function types triggers warning), perhaps a `NumbaExperimentalFeatureWarning` or similar.
* Longer term goal: compilation should derive from more in depth static analysis.

Merge conditions:
* @pearu are you able to commit to providing support for fixing parts of this that may need alterations and also to respond to user bugs? Thanks.

pearu commented:
@stuartarchibald @sklam , yes, I would be happy to continue improving the feature and fix any related bugs.

stuartarchibald commented:
Thanks @pearu, that's great!

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+objects (that are collected in the input argument ``funcs``).
+
+Currently, first-class function objects can be Numba ``cfunc`` compiled
+functions, ``jit`` compiled functions, and objects that implement
```
```suggestion
functions, ``jit`` compiled functions, and objects that implement the
```

pearu replied:
Fixed.

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+
+Currently, first-class function objects can be Numba ``cfunc`` compiled
+functions, ``jit`` compiled functions, and objects that implement
+Wrapper Address Protocol (WAP) with the following restrictions:
```
```suggestion
Wrapper Address Protocol (WAP, see below) with the following restrictions:
```

pearu replied:
Fixed.

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+++++++++++++++++++++++++++++++
+
+Wrapper Address Protocol provides an API for making any Python object
+as first-class function for Numba jit compiled functions. This assumes
```
```suggestion
a first-class function for Numba jit compiled functions. This assumes
```

pearu replied:
Fixed.

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+that the Python object represents a compiled function that can be
+called via its memory address (function pointer value) from Numba jit
+compiled functions. Such the so-called WAP objects must define the
+following two methods:
```
```suggestion
compiled functions. The so-called WAP objects must define the
following two methods:
```

pearu replied:
Fixed.

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+.. method:: __wrapper_address__(self) -> int
+
+            Return the memory address of a first-class function. This
+            method is used when Numba jit compiled function tries to
```
```suggestion
            method is used when a Numba jit compiled function tries to
```

pearu replied:
Fixed.

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+
+            Return the signature of the given first-class
+            function. This method is used when passing in the given
+            WAP instance to Numba jit compiled function.
```
```suggestion
            WAP instance to a Numba jit compiled function.
```

pearu replied:
Fixed.

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+            WAP instance to Numba jit compiled function.
+
+In addition, to make WAP objects callable from Numba jit compiled
+function in object mode, the WAP object must also implement the
```
```suggestion
functions in object mode, the WAP object must also implement the
```

stuartarchibald replied:
Also, "must" should perhaps be "may", the example given directly below does not implement it.

pearu replied:
OK. Just to be clear, the "must" applies when one needs to use the functions in *object mode*. The example works in *nopython mode*.

pearu replied:
Rephrased the sentence as follows
```
In addition, the WAP object may implement the ``__call__``
method. This is necessary when calling WAP objects from Numba
:term:`JIT` compiled functions in :term:`object mode`.
```

stuartarchibald replied:
looks good, thanks

stuartarchibald commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+
+As an example, let us call the standard math library function ``cos``
+within a Numba jit compiled function. The memory address of ``cos`` can
+be established after loading the math library and using ctypes
```
```suggestion
be established after loading the math library and using the ``ctypes``
```

pearu replied:
Fixed.

pearu commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+   The feature of considering functions as first-class type objects is
+   under development.
+
+While functions are often considered as certain transformations of
```
Fixed, thanks!

pearu commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+   under development.
+
+While functions are often considered as certain transformations of
+input arguments to output values then within Numba jit compiled
```
Fixed.

pearu commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+input arguments to output values then within Numba jit compiled
+functions, the functions can also be considered as objects, that is,
+functions can be passed around as arguments or return values, or used
+as items in sequences, in addition of being callable.
```
Fixed.

pearu commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+functions can be passed around as arguments or return values, or used
+as items in sequences, in addition of being callable.
+
+The first-class function support is enabled for all Numba :term:`JIT`
```
Fixed.

pearu commented on docs/source/reference/types.rst:
```diff
@@ -100,6 +100,122 @@ the beginning or the end of the index specification::
+as items in sequences, in addition of being callable.
+
+The first-class function support is enabled for all Numba :term:`JIT`
+compiled functions and Numba ``cfunc``-decorated functions except when
```
Fixed.

stuartarchibald commented:
Fixes to the docs in 984a24b look good, thanks @pearu .

sklam commented:
https://github.com/numba/numba/pull/5441 should be merged immediately after merging this
